platforms/jvm/ear/src/main/java/org/gradle/plugins/ear/Ear.java

```diff
@@ -108,10 +103,15 @@ public Ear() {
                     getTemporaryDirFactory(),
                     descriptorFileName,
                     action(file -> outputChangeListener.invalidateCachesFor(singleton(file.getAbsolutePath()))),
-                    action(outputStream ->
-                        // delay obtaining contents to account for descriptor changes
-                        // (for instance, due to modules discovered)
-                        descriptor.writeTo(new OutputStreamWriter(outputStream))
+                    action(outputStream -> {
+                            // delay obtaining contents to account for descriptor changes
+                            // (for instance, due to modules discovered)
+                            // We copy the descriptor since properties are already finalized at that moment with cc
+                            DeploymentDescriptor copyDescriptor = getObjectFactory().newInstance(DefaultDeploymentDescriptor.class).copyFrom(descriptor);
+                            copyDescriptor.getModules().addAll(topLevelModules.get());
+                            copyDescriptor.getLibraryDirectory().set(descriptor.getLibraryDirectory().orElse(getLibDirName()));
+                            copyDescriptor.writeTo(new OutputStreamWriter(outputStream));
```

Comment:
💭 We do this copy logic because properties can be finalized at that moment with cc

Reply:
:x: Why is it a problem that the properties are finalized? IIUC, this is about having to call `setLibraryDirectory()`, right? Why do we have to call this at execution time? Can't we set that as a convention instead?

Reply:
> Why is it a problem that the properties are finalized?

The main problem why we do it here is because we want to add these top level modules, so this line:
```
copyDescriptor.getModules().addAll(topLevelModules.get())
``` 

I have to refresh my memory why `setLibraryDirectory()` is a problem here